feat: Add dynamic parameter config source support for DAG params

chen0427ok opened the pull request:
## Description

Closes #56632

This PR adds support for loading DAG parameter options dynamically from external configuration files (INI, JSON, YAML) with filtering capabilities.

## Changes

### New Files
- **`airflow-core/src/airflow/utils/param_config_loader.py`**: Utility functions to load options from config files
  - `load_options_from_ini()` - Load from INI files with filtering
  - `load_options_from_json()` - Load from JSON files with filtering
  - `load_options_from_yaml()` - Load from YAML files with filtering
  - All support filter conditions and custom key fields

### Modified Files
- **`task-sdk/src/airflow/sdk/definitions/param.py`**: Extended `Param` class
  - Added `config_source` parameter to `__init__()`
  - Added `_load_options_from_config()` method to load options from external files
  - Updated `__copy__()`, `serialize()`, and `deserialize()` methods
  - Automatically populates `schema["enum"]` from config files at DAG parse time

### Test Files
- **`airflow-core/tests/unit/utils/test_param_config_loader.py`**: 18 unit tests for config loader
- **`task-sdk/tests/task_sdk/definitions/test_param.py`**: 4 new tests for Param class config_source feature

### Documentation
- **`airflow-core/newsfragments/56632.feature.rst`**: Release note for this feature

## Usage Example

**Config file** (`/path/to/interfaces.ini`):
```ini
[InterfaceA]
TYPE = Script
DESCRIPTION = Script based interface A

[InterfaceB]
TYPE = EBICS
DESCRIPTION = Electronic Banking interface

[InterfaceC]
TYPE = Script
DESCRIPTION = Another script interface

DAG definition:
from airflow import DAG
from airflow.sdk import Param
from datetime import datetime

with DAG(
    dag_id="example_dynamic_params",
    start_date=datetime(2024, 1, 1),
    schedule=None,
    params={
        "interface_name": Param(
            default="InterfaceA",
            config_source={
                "file": "/path/to/interfaces.ini",
                "filter": {"TYPE": "Script"},  # Only show Script type interfaces
                "key_field": "section",
            },
            description="Select an interface (dynamically loaded)",
        ),
    },
) as dag:
    # ... tasks

Result in UI:
- Renders a dropdown with only InterfaceA and InterfaceC (filtered by TYPE = Script)
- No frontend code changes needed - uses existing enum schema mechanism

Supported Config Formats

INI Files

- Filter by section properties
- Extract section names or specific field values
- Example: filter interfaces by type, environment, etc.

JSON Files

- Support array of objects
- Filter by object properties
- Extract specific field as option value

YAML Files

- Support array of objects
- Filter by object properties
- Extract specific field as option value

Testing

✅ All 22 unit tests passing:
- 18 tests for param_config_loader.py covering all formats and edge cases
- 4 tests for Param class config_source integration

✅ Manual testing in Airflow UI:
- Created test DAG with config_source parameter
- Verified dropdown rendering with filtered options
- Verified task execution with selected parameter values
- Tested with all three config formats (INI, JSON, YAML)

✅ Static checks:
- All prek hooks passed
- ruff formatting/linting passed
- mypy type checking passed

Test Commands

# Test config loader
uv run pytest airflow-core/tests/unit/utils/test_param_config_loader.py -v

# Test Param class integration
uv run pytest task-sdk/tests/task_sdk/definitions/test_param.py -k "config_source" -v

Implementation Notes

- Config files are loaded at DAG parse time, not at UI render time
- Options are stored in the enum field of the parameter's JSON schema
- If config loading fails (file not found, parsing error), a warning is logged and the parameter falls back to a regular string field
- No database schema changes required
- No UI code changes required - reuses existing enum dropdown rendering
- Backward compatible - existing DAGs without config_source continue to work unchanged

Follow-up Work

This PR provides the foundation for dynamic parameter options. Potential future enhancements:
- Support for database queries as config sources
- Support for API endpoints as config sources
- Caching mechanism for frequently accessed config files
- UI for managing config files

Checklist

- Feature implementation completed
- Unit tests added and passing (22/22)
- Manual testing in UI completed
- Static checks passing
- Newsfragment added
- Code follows Airflow coding standards
- Documentation in code (docstrings)
- Backward compatible

---

chen0427ok committed:
feat: Add dynamic parameter config source support

Implement config_source parameter for Param class to load dropdown
options from external configuration files (INI, JSON, YAML).

This addresses issue #56632 by allowing DAG authors to define parameter
options dynamically from configuration files with optional filtering.

Changes:
- Add param_config_loader.py with support for INI/JSON/YAML files
- Extend Param class to accept config_source parameter
- Options are loaded at DAG parse time and added to schema as enum
- Includes comprehensive docstrings and error handling

Example usage:
    Param(
        default='InterfaceA',
        config_source={
            'file': 'config/interfaces.ini',
            'filter': {'TYPE': 'Script'},
            'key_field': 'section'
        }
    )

chen0427ok committed:
test: Add unit tests for config_source parameter feature

Add comprehensive unit tests for the new config_source functionality:

- Test load_options_from_ini with filtering and key selection
- Test load_options_from_json with filtering
- Test load_options_from_yaml with filtering
- Test Param class integration with config_source
- Test error handling and serialization

All tests use pytest fixtures and follow Airflow testing conventions.

chen0427ok committed:
docs: Add newsfragment for config_source feature

chen0427ok committed:
refactor: Move param_config_loader entirely to task-sdk

Move config loader from airflow-core to task-sdk to maintain
task-sdk independence from airflow-core as requested by reviewer.

- Move param_config_loader.py to task-sdk/src/airflow/sdk/definitions/
- Move test_param_config_loader.py to task-sdk/tests/task_sdk/definitions/
- Update imports in param.py to use task-sdk location
- Remove try-except ImportError since loader is now in same package

All tests passing (18 config loader tests + 4 Param integration tests)

jason810496 commented on task-sdk/src/airflow/sdk/definitions/param_config_loader.py:
```diff
@@ -0,0 +1,243 @@
+        options = load_options_from_ini("config/interfaces.ini", key_field="DESCRIPTION")
+        # Returns: ["Script interface A", "Electronic banking interface"]
+    """
+    file_path = Path(file_path)
+    if not file_path.exists():
+        raise FileNotFoundError(f"Configuration file not found: {file_path}")
```
Would it be better to consolidate the three method as `load_options_from_file` and provide `extension: Literal["ini", "json", "yaml"]` parameter?

Since the validation of file path and filter of `filter_conditions` and `key_field` are duplicated.

chen0427ok replied:
Thanks for the review and the great suggestion! I agree that consolidating them into a single load_options_from_file function will reduce code duplication and make it easier to maintain. I'll refactor the code to merge the validation and filtering logic.

jason810496 commented on task-sdk/src/airflow/sdk/definitions/param.py:
```diff
@@ -56,6 +72,7 @@ def __init__(
         default: Any = NOTSET,
         description: str | None = None,
         source: Literal["dag", "task"] | None = None,
+        config_source: dict[str, Any] | None = None,
```
How about adding a `TypeDict` for `config_source`? As `dict[str, Any]` seems to be too ambiguous for user.

chen0427ok replied:
That's a great suggestion. Using TypedDict will definitely make the expected structure much clearer for users and remove the ambiguity of dict[str, Any]. I'll define a ConfigSource TypedDict and apply it.

jason810496 commented on task-sdk/src/airflow/sdk/definitions/param.py:
```diff
@@ -64,13 +81,67 @@ def __init__(
+            key_field = key_field or "name"
+            return load_options_from_yaml(file_path, filter_conditions, key_field)
+        raise ValueError(
+            f"Unsupported config file type: {file_path}. Supported types are: .ini, .json, .yaml, .yml"
```
Then we could just pass the file extension to `filter_conditions`.

chen0427ok replied:
Thanks for the detailed review! I've addressed all your feedback:

  1. Consolidated the three loader functions into `load_options_from_file()` with an `extension` parameter

  2. Added `ConfigSource` TypedDict for better type hints and IDE support

  3. Removed unnecessary try-catch for yaml import (already in airflow-core dependencies)

  4. All tests passing (19 config loader tests + 4 Param integration tests)

  The refactoring follows the DRY principle and makes the code much more maintainable. Let me know if you need any other changes!

jason810496 commented on task-sdk/src/airflow/sdk/definitions/param.py:
```diff
@@ -64,13 +81,67 @@ def __init__(
+                options = self._load_options_from_config(config_source)
+                if options:
+                    self.schema["enum"] = options
+            except Exception as e:
```
We might need to add a new exception in `task-sdk/src/airflow/sdk/exceptions.py` instead of raising built-in `Exception`.

jason810496 commented on task-sdk/src/airflow/sdk/definitions/param_config_loader.py:
```diff
@@ -0,0 +1,243 @@
+    try:
+        import yaml
+    except ImportError as e:
+        raise ImportError("PyYAML is required to load YAML files. Install it with: pip install PyYAML") from e
```
It seems we don't need to have the try-catch for importing `yaml` package as task-sdk is depends on airflow-core

https://github.com/apache/airflow/blob/7c6fcecc535de447482b18c85fb98235655e745d/task-sdk/pyproject.toml#L52-L53

and `pyyaml` is already included in airflow-core.

https://github.com/apache/airflow/blob/7c6fcecc535de447482b18c85fb98235655e745d/airflow-core/pyproject.toml#L157-L158

chen0427ok replied:
Thanks for pointing that out! I didn't realize pyyaml was already a dependency in airflow-core. I'll remove the try-catch block and import yaml directly.

jason810496 reviewed (commented):
Nice! Thanks for the PR!

chen0427ok committed:
refactor: Consolidate config loaders and add TypedDict for config_source

Address reviewer feedback from jason810496:

1. Consolidate three separate loader functions into single load_options_from_file()
   - Eliminates duplicated validation logic for file_path, filter_conditions, key_field
   - Uses 'extension' parameter to specify file format (ini/json/yaml)
   - Centralizes common logic in helper functions

2. Add TypedDict definition for config_source parameter
   - Provides clear type hints for IDE autocomplete
   - Documents expected structure: file, filter, key_field
   - Improves developer experience and code documentation

3. Remove try-catch for yaml import
   - task-sdk depends on airflow-core which includes pyyaml
   - No need for conditional import handling

4. Update all tests to use new unified function
   - 19 tests for load_options_from_file covering all formats
   - All tests passing (19 config loader + 4 Param integration)

This refactoring follows DRY principle and makes the code more maintainable.

jason810496 reviewed (commented):
After I read the issue again, I thought the purpose of the issue would be like:

1. The administrator are able to define the param config file
```json
[
    {
        "name": "InterfaceA",
        "type": "dev",
        "params": {
            "number_param": 3,
            "text_param": "Hello World!",
            "bool_param": false
        }
    },
    {
        "name": "InterfaceB",
        "type": "prod",
        "params": {
            "list_param": ["one", "two", "three", "actually one value is made per line"],
            "dict_param": {"key": "value"}
        }
    }
]
```

2. The Dag users are able to specific which param config to at Dag level

```python
with DAG(
    "finance_prod_dag",
    params={
        "config_source": {
            "source": "config/param_config.json",
            "filter": {"type": "prod"},
            "key_field": "params"
        }
    },
) as dag:
    ...

```

From the above example, the Dag should actually be rendered as 


```python
with DAG(
    "finance_prod_dag",
    params={
        "list_param": ["one", "two", "three", "actually one value is made per line"],
        "dict_param": {"key": "value"}
    }
) as dag:
    ...

```

so that users are able to reuse the same Dag Param from config file. 

@potiuk May I ask if we’re aligned on this, and what you think of the proposed user‑facing change? I’m not entirely sure I’ve captured the issue context correctly. Thanks!
